## car price tracker
2023/01/14
12:23


## import data

In [10]:
import csv


f = open("C:/Users/OWO/Documents/AA_CODE/@Projects/Projects/ai_0002_lr_car price track/train_data.csv")
read = csv.reader(f)

#getting data from the csv file
data = []
for a in read:
  data.append(a)
  
#print(data)
x_train  = data[:-4][1:]
y_train  = data[:][-1]
#print(x_train)

#deleting the first row because thats the legend
del data[0]


car_year_test = []
car_trans_test = []#transmisstion 1 == manual //  3 ==  dual clutch // 
car_milege = []
car_fueltype =[]# 4 == Petrol // 0 == premium gasoline // 2 == gasoline 
car_tax = []
car_mpg = []# mile per gallon
car_enginesize = []
car_price = []



templist = []

#print(data)

for count in range(0,17000):
  
  templist.append(float(data[count][3]))
  car_year_test.append(templist)
  templist = []
  templist.append(float(data[count][4]))
  car_trans_test.append(templist)
  templist = []
  templist.append(float(data[count][5]))
  car_milege.append(templist)
  templist = []
  templist.append(float(data[count][6]))
  car_fueltype.append(templist)
  templist = []
  templist.append(float(data[count][7]))
  car_tax.append(templist)
  templist = []
  templist.append(float(data[count][8]))
  car_mpg.append(templist)
  templist = []  
  templist.append(float(data[count][9]))
  car_enginesize.append(templist)
  templist = []
  templist.append(float(data[count][10]))
  car_price.append(templist)
  templist = []   
  

f.close()
#print(pricelist)
print("total data count : "+str(len(car_price)))


total data count : 17000


In [12]:
##Hardwere
import torch
if torch.cuda.is_available() == True:
  device = 'cuda'
  templist = [1,2,3]
  templist = torch.FloatTensor(templist).to(device)
  print("Cude torch working : ",end="")
  print(templist.is_cuda)
  print("current device no. : ",end="")
  print(torch.cuda.current_device())
  print("GPU device count : ",end="")
  print(torch.cuda.device_count())
  print("GPU name : ",end="")
  print(torch.cuda.get_device_name(0))
  print("device : ",device)
else:
  print("cant use gpu , activating cpu")
  device = 'cpu'

cant use gpu , activating cpu


## ML Method no.1

Matmul

In [4]:
import torch
import torch.optim as optim
torch.manual_seed(1)


##Hardware
print("GPU Status ",end="")
print("Working Well") if torch.cuda.is_available() else print("Not good, Activating CPU calculation")

device = 'cuda' if torch.cuda.is_available() else 'cpu'



data_torch = []

car_year_test = torch.FloatTensor(car_year_test).to(device)
car_trans_test = torch.FloatTensor(car_trans_test).to(device)
car_milege = torch.FloatTensor(car_milege).to(device)
car_fueltype = torch.FloatTensor(car_fueltype).to(device)
car_tax = torch.FloatTensor(car_tax).to(device)
car_mpg = torch.FloatTensor(car_mpg).to(device)
car_enginesize = torch.FloatTensor(car_enginesize).to(device)
car_price = torch.FloatTensor(car_price).to(device)

data_torch = torch.cat([car_year_test,car_trans_test,car_milege,car_fueltype,car_tax,car_mpg,car_enginesize],dim = 1)

lr_list = [1,1e-2,1e-3,1e-4,1e-5,1e-6,1e-7,1e-8,1e-9,1e-10,1e-11,1e-12,1e-13,1e-14,1e-15]
#lr_list = [1e-5]
for LR in lr_list:
  
  W = torch.zeros([7,1],requires_grad=True, device=device)
  b = torch.zeros(1,requires_grad=True, device=device)
  
  optimizer = optim.SGD( [W] , lr = LR)
  nb_epochs = 1000

  for epoch in range(nb_epochs + 1):
    # H(x) 계산
    # Matrix 연산!!
    hypothesis = torch.matmul(data_torch,W) + b # or .mm or @
    
    # cost 계산
    cost = torch.mean(torch.abs(hypothesis - car_price))

    # cost로 H(x) 개선
    optimizer.zero_grad()
    cost.backward()
    optimizer.step()
    """
    if epoch % 100000 == 0:
        print('Epoch {:4d}/{} Cost: {:.6f}'.format(
            epoch, nb_epochs, cost.item()
        ))
        print(LR)
        print(W)
        print(b)
        """


  print("learing rate : %.2e COST : %f"%(LR,cost.item()))
  print("W : ",end="")
  print(W)
  print("b : ",end="")
  print(b)


GPU Status Not good, Activating CPU calculation
learing rate : 1.00e+00 COST : 16824.837891
W : tensor([[2.0169e+03],
        [1.1870e+00],
        [2.3814e+04],
        [2.2851e+00],
        [1.1873e+02],
        [5.6451e+01],
        [1.6721e+00]], requires_grad=True)
b : tensor([0.], requires_grad=True)
learing rate : 1.00e-02 COST : 63126.582031
W : tensor([[2.4853e+01],
        [1.8357e-02],
        [2.3579e+02],
        [1.1579e-02],
        [1.4585e+00],
        [6.3548e-01],
        [3.4688e-02]], requires_grad=True)
b : tensor([0.], requires_grad=True)
learing rate : 1.00e-03 COST : 156055.796875
W : tensor([[ 1.3016e+01],
        [ 2.4452e-02],
        [ 1.7019e+01],
        [-1.2022e-02],
        [ 9.5239e-01],
        [ 1.4236e-01],
        [ 3.2636e-02]], requires_grad=True)
b : tensor([0.], requires_grad=True)
learing rate : 1.00e-04 COST : 27635.173828
W : tensor([[ 1.1740e+01],
        [ 1.4321e-02],
        [ 8.9641e-01],
        [-3.5021e-03],
        [ 9.8437e-01],
 

Epoch 1000000/1000000 Cost: 5920.548828
1e-05
tensor([[  7.3967],
        [  1.6980],
        [ -0.1158],
        [ -8.0022],
        [ 34.3764],
        [-15.1122],
        [  3.5023]], device='cuda:0', requires_grad=True)
tensor([0.], device='cuda:0', requires_grad=True)

## finding optim

## Result : 
learing rate : 1.00e+00 COST : 16824.837891
W : tensor([[2.0170e+03],
        [1.1871e+00],
        [2.3814e+04],
        [2.2849e+00],
        [1.1872e+02],
        [5.6452e+01],
        [1.6722e+00]], device='cuda:0', requires_grad=True)
learing rate : 1.00e-02 COST : 95406.703125
W : tensor([[ 2.5274e+01],
        [ 1.2596e-01],
        [ 2.3443e+02],
        [-1.2783e+00],
        [-2.0781e+00],
        [-2.2283e+00],
        [ 1.2126e+00]], device='cuda:0', requires_grad=True)
learing rate : 1.00e-03 COST : 192555.546875
W : tensor([[ 13.4784],
        [  1.4357],
        [ 15.4722],
        [ -2.3972],
        [  5.1327],
        [-17.8910],
        [  2.1180]], device='cuda:0', requires_grad=True)
learing rate : 1.00e-04 COST : 29200.712891
W : tensor([[ 11.6786],
        [  0.7139],
        [  0.8168],
        [ -1.8871],
        [ 11.4326],
        [-11.6211],
        [  1.1623]], device='cuda:0', requires_grad=True)
## OPTIM LR : 1e-5
learing rate : 1.00e-05 COST : 6063.766113
W : tensor([[ 8.2146],
        [ 0.1875],
        [-0.1302],
        [-0.7425],
        [14.9173],
        [-2.6845],
        [ 0.3787]], device='cuda:0', requires_grad=True)
learing rate : 1.00e-06 COST : 6251.767578
W : tensor([[ 8.9776],
        [ 0.0229],
        [-0.1392],
        [-0.0586],
        [ 2.3025],
        [-0.0705],
        [ 0.0430]], device='cuda:0', requires_grad=True)
learing rate : 1.00e-07 COST : 6546.291016
W : tensor([[ 7.3240e+00],
        [ 5.5396e-03],
        [-6.8594e-02],
        [ 5.5401e-03],
        [ 6.2660e-01],
        [ 1.7800e-01],
        [ 7.9652e-03]], device='cuda:0', requires_grad=True)
learing rate : 1.00e-08 COST : 11479.930664
W : tensor([[1.1070e+00],
        [7.4917e-04],
        [2.6725e-01],
        [1.3711e-03],
        [8.1947e-02],
        [2.9012e-02],
        [9.6261e-04]], device='cuda:0', requires_grad=True)
learing rate : 1.00e-09 COST : 12559.625000
W : tensor([[1.2567e-01],
        [8.2345e-05],
        [3.2335e-01],
        [1.5415e-04],
        [8.7875e-03],
        [3.3620e-03],
        [1.0737e-04]], device='cuda:0', requires_grad=True)
learing rate : 1.00e-10 COST : 13475.258789
W : tensor([[1.8587e-02],
        [1.1162e-05],
        [1.7761e-01],
        [2.1309e-05],
        [1.1109e-03],
        [5.1942e-04],
        [1.5431e-05]], device='cuda:0', requires_grad=True)
learing rate : 1.00e-11 COST : 16261.408203
W : tensor([[2.0146e-03],
        [1.1861e-06],
        [2.3649e-02],
        [2.2816e-06],
        [1.1842e-04],
        [5.6409e-05],
        [1.6697e-06]], device='cuda:0', requires_grad=True)
learing rate : 1.00e-12 COST : 16767.703125
W : tensor([[2.0186e-04],
        [1.1867e-07],
        [2.3821e-03],
        [2.2871e-07],
        [1.1886e-05],
        [5.6418e-06],
        [1.6728e-07]], device='cuda:0', requires_grad=True)
learing rate : 1.00e-13 COST : 16819.121094
W : tensor([[2.0187e-05],
        [1.1879e-08],
        [2.3835e-04],
        [2.2868e-08],
        [1.1867e-06],
        [5.6459e-07],
        [1.6713e-08]], device='cuda:0', requires_grad=True)
learing rate : 1.00e-14 COST : 16824.267578
W : tensor([[2.0154e-06],
        [1.1877e-09],
        [2.3826e-05],
        [2.2866e-09],
        [1.1867e-07],
        [5.6483e-08],
        [1.6736e-09]], device='cuda:0', requires_grad=True)
learing rate : 1.00e-15 COST : 16824.781250
W : tensor([[2.0178e-07],
        [1.1868e-10],
        [2.3806e-06],
        [2.2870e-10],
        [1.1879e-08],
        [5.6409e-09],
        [1.6702e-10]], device='cuda:0', requires_grad=True)


## ML Method 2

its the same thing (in theory) 

In [8]:
import torch
import torch.optim as optim

car_year_test = torch.FloatTensor(car_year_test)
car_trans_test = torch.FloatTensor(car_trans_test)
car_milege = torch.FloatTensor(car_milege)
car_fueltype = torch.FloatTensor(car_fueltype)
car_tax = torch.FloatTensor(car_tax)
car_mpg = torch.FloatTensor(car_mpg)
car_enginesize = torch.FloatTensor(car_enginesize)
car_price = torch.FloatTensor(car_price)







lr_list = [1e-3,1e-4,1e-5,1e-6,5e-7,2e-8,1e-9,1e-10,1e-11,1e-12]
#lr_list = [9.7e-10]
for LR in lr_list:
    
  w1 = torch.ones(1, requires_grad=True)
  w2 = torch.ones(1,requires_grad=True)
  w3 = torch.ones(1,requires_grad=True)
  w4 = torch.ones(1,requires_grad=True)
  w5 = torch.ones(1,requires_grad=True)
  w6 = torch.ones(1,requires_grad=True)
  w7 = torch.ones(1,requires_grad=True)#car engine
  b  = torch.ones(1,requires_grad=True)


  optimizer = optim.SGD([w1,w2,w3,w4,w5,w6,w7,b], lr=LR) #learning rate

  nb_epochs = 1000

  for epoch in range(nb_epochs + 1):
    #hypothesis calcuation
    hypothesis = car_year_test * w1 + car_trans_test * w2 + car_milege * w3 + car_fueltype * w4 + car_tax * w5 + car_mpg * w6 + car_enginesize * w7 + b
    
    #cost calculation
    cost = torch.mean(torch.abs(hypothesis - car_price))
    
    #impove H(x) with cost
    optimizer.zero_grad()
    cost.backward()
    optimizer.step()
    
    if epoch%1000 == 0:
      print("LR = %.2e Epoch : %d w1 : %.1f w2 : %.1f w3 : %.1f w4 : %.1f w5 : %.1f w6 : %.1f w7 : %.1f b : %.1f Cost : %.0f"%(LR,epoch,w1.item(),w2.item(),w3.item(),w4.item(),w5.item(),w6.item(),w7.item(),b,cost.item()))
      
print("LR = %.2e Epoch : %d w1 : %.1f w2 : %.1f w3 : %.1f w4 : %.1f w5 : %.1f w6 : %.1f w7 : %.1f b : %.1f Cost : %.0f"%(LR,nb_epochs,w1.item(),w2.item(),w3.item(),w4.item(),w5.item(),w6.item(),w7.item(),b,cost.item()))


LR = 1.00e-03 Epoch : 0 w1 : 0.5 w2 : 1.0 w3 : -17.7 w4 : 1.0 w5 : 1.0 w6 : 1.0 w7 : 1.0 b : 1.0 Cost : 20898
LR = 1.00e-03 Epoch : 1000 w1 : 11.6 w2 : 1.0 w3 : -9.8 w4 : 1.0 w5 : 1.8 w6 : 1.1 w7 : 1.0 b : 1.0 Cost : 345648
LR = 1.00e-04 Epoch : 0 w1 : 1.0 w2 : 1.0 w3 : -0.9 w4 : 1.0 w5 : 1.0 w6 : 1.0 w7 : 1.0 b : 1.0 Cost : 20898
LR = 1.00e-04 Epoch : 1000 w1 : 11.5 w2 : 1.0 w3 : -1.4 w4 : 1.0 w5 : 1.9 w6 : 1.1 w7 : 1.0 b : 1.0 Cost : 30394
LR = 1.00e-05 Epoch : 0 w1 : 1.0 w2 : 1.0 w3 : 0.8 w4 : 1.0 w5 : 1.0 w6 : 1.0 w7 : 1.0 b : 1.0 Cost : 20898
LR = 1.00e-05 Epoch : 1000 w1 : 7.5 w2 : 1.0 w3 : -0.1 w4 : 1.0 w5 : 1.6 w6 : 1.2 w7 : 1.0 b : 1.0 Cost : 6455
LR = 1.00e-06 Epoch : 0 w1 : 1.0 w2 : 1.0 w3 : 1.0 w4 : 1.0 w5 : 1.0 w6 : 1.0 w7 : 1.0 b : 1.0 Cost : 20898
LR = 1.00e-06 Epoch : 1000 w1 : 2.0 w2 : 1.0 w3 : 0.2 w4 : 1.0 w5 : 1.1 w6 : 1.0 w7 : 1.0 b : 1.0 Cost : 10406
LR = 5.00e-07 Epoch : 0 w1 : 1.0 w2 : 1.0 w3 : 1.0 w4 : 1.0 w5 : 1.0 w6 : 1.0 w7 : 1.0 b : 1.0 Cost : 20898
LR = 5.0

## finding optim

my best optimized w




Epoch 1000000/1000000 Cost: 5920.548828
1e-05
tensor([[  7.3967],
        [  1.6980],
        [ -0.1158],
        [ -8.0022],
        [ 34.3764],
        [-15.1122],
        [  3.5023]], device='cuda:0', requires_grad=True)
tensor([0.], device='cuda:0', requires_grad=True)

## input NEW test data

In [49]:
import csv


f = open("C:/Users/OWO/Documents/AA_CODE/@Projects/Projects/ai_0002_lr_car price track/train_data.csv")
read = csv.reader(f)

#getting data from the csv file
data = []
for a in read:
  data.append(a)
#deleting the first row because thats the legend
del data[0]


new_car_year_test = []
new_car_trans_test = []#transmisstion
new_car_milege = []
new_car_fueltype =[]# 4 == Petrol // 0 == premium gasoline // 2 == gasoline 
new_car_tax = []
new_car_mpg = []# mile per gallon
new_car_enginesize = []
new_car_price = []



templist = []

#print(data)

for count in range(17000,17084):
  
  templist.append(float(data[count][3]))
  new_car_year_test.append(templist)
  templist = []
  templist.append(float(data[count][4]))
  new_car_trans_test.append(templist)
  templist = []
  templist.append(float(data[count][5]))
  new_car_milege.append(templist)
  templist = []
  templist.append(float(data[count][6]))
  new_car_fueltype.append(templist)
  templist = []
  templist.append(float(data[count][7]))
  new_car_tax.append(templist)
  templist = []
  templist.append(float(data[count][8]))
  new_car_mpg.append(templist)
  templist = []  
  templist.append(float(data[count][9]))
  new_car_enginesize.append(templist)
  templist = []
  templist.append(float(data[count][10]))
  new_car_price.append(templist)
  templist = []   
  

f.close()
#print(pricelist)
print("total data count : "+str(len(car_price)))
test_count = len(car_price)

total data count : 17000


## Test cost

In [50]:
#LR = 1e-5 Epoch 1000000/1000000 Cost: 5920.548828
w1 = 7.3967
w2 = 1.698
w3 = -0.1158
w4 = -8.0022
w5 = 34.3764
w6 = -15.1122
w7 = 3.5023
b = 0


cost_sum = 0
test_count  = len(new_car_year_test)

for row in range(0,test_count):
  hypothesis = w1 * new_car_year_test[row][0] + w2 * new_car_trans_test[row][0] + w3 * new_car_milege[row][0] + w4 * new_car_fueltype[row][0] + w5 * new_car_tax[row][0] + w6 * new_car_mpg[row][0] + w7 * new_car_enginesize[row][0] + b
  cost_sum = hypothesis - car_price[row][0] + cost_sum

print("avrage diffrence : ",end='')
print("%.0f"%(cost_sum/test_count))

avrage diffrence : -4532


## the End code

In [51]:
#Data
w1 = 7.3967
w2 = 1.698
w3 = -0.1158
w4 = -8.0022
w5 = 34.3764
w6 = -15.1122
w7 = 3.5023
b = 0


print("hello this is the Car Price Predictor")
print("type in the data then you will recive the price by its data")


new_car_year = float(input("type in the car's production year : "))

new_car_trans = 3 #default :  auto == 3
trans = input("type in the transmisstion : \n if manual type m . if auto type a")
if trans == 'm':#if manual
  new_car_trans = 1

try:
  new_car_milage = float(input("type in the mileage of the car : \nif you want to type the kilometers, type in k"))
except:
  new_car_milage =  0.62137119 * float(input("type in the kilometers of the car : "))
  
new_car_fueltype = 4#default petrol
fuletype =  input("type in the fuel type of the car\n if its petrol type p, if its premium gasoline type pr , if its diesel type in  d : ")# 4 == Petrol // 0 == premium gasoline // 2 == disel 
if fuletype == 'pr':
  new_car_fueltype = 0
elif fuletype == 'd':
  new_car_fueltype = 2
  
new_car_tax = float(input("input the tax of the car you pay a year (in dollors) : "))
new_car_mpg = float(input("input the miles per gallon of the car : "))
new_car_enginesize = float(input("input the car enginesize by liters"))

print(" the AI is guessing it would be : ",end="")
print("$ %.0f "%(new_car_year * w1 + new_car_trans * w2 + new_car_milage * w3 + new_car_fueltype * w4 + new_car_tax * w5 + new_car_mpg * w6 + w7 * new_car_enginesize + b)
)

hello this is the Car Price Predictor
type in the data then you will recive the price by its data
 the AI is guessing it would be : $ 18937 


## End of project

update count : 1

2023/01/16 1325i -----------

I solved the nan probelm, the cost was getting to high and it made the cost infinite.
So I used aboslute rather than **2
And also I used GPU




2023/01/15 1225i---------------------- 
well I change my way to calculate in the last moment (the ML mothod 1)
it makes stuff much more simple. why didnt I thought of that at from the start!
And the mytsery thing was when the LR was just fine with 1e-9, but after when I made the ML mothod 1, 1e-9 didnt work... (I didnt find out the reason yet) But I just keeped the model (the W) with the previous one (the one with lr : 1e-9)

the AI performance is not that great.. as you see , at the TEST cost, it has an avrage -$1230 diffrence.
It was a fun project thogh. I leanred a lot.
its 2023/01/15 1225i
I think it took about 6 hours. Half of them was research.

the data source : 
https://data.kma.go.kr/climate/StatisticsDivision/selectStatisticsDivision.do?pgmNo=158
https://www.kamis.or.kr/customer/price/retail/period.do?action=monthly&yyyy=2023&period=28&countycode=&itemcategorycode=100&itemcode=111&kindcode=&productrankcode=&convert_kg_yn=N

